### ISSUE TITLE
Generate model associations (belongsTo, hasMany, belongsToMany) from foreign keys

### ISSUE THREAD

**Issue by eighteight** [sequelize/sequelize-auto#34]
We know it is possible to generate the model associations above for the [sequelize](http://docs.sequelizejs.com/en/latest/docs/associations/) to use, similar to how sequelize-pg-generator does.

**Comment by Sawtaytoes** [comment]
I'd like this as well. I'm using MySQL and GraphQL. I noticed after running this tool, it put in all the references necessary, but didn't associate them with each other such that I could call something like "employees.getUsers()".

**Comment by Jarvie8176** [comment]
need to add classMethods for associations.

Here are some examples:

https://github.com/sequelize/express-example/tree/master/models

I added those methods manually to the model files generated and it worked.

**Comment by danny-larsen** [comment]
Does anybody know if there is any way to generate these associate functions?

We have a fairly large database and it would help to have these generated automatically.

**Comment by ahelord** [comment]
Good afternoon, I recently use sequelize I want to know if associations (hasMany, hasOne, etc.) between models can be generated at this time with sequelize-auto

**Comment by NobleUplift** [comment]
Is it possible to inject associations into Sequelize after importing the model definitions generated by Sequelize Auto? This is the code I've been trying to use but it hasn't worked:

```
			console.log("Length of models: " + Object.keys(instance.sequelize.models).length)
			for (var objectName in instance.sequelize.models) {
				var model = instance.sequelize.models[objectName];
				var columns = instance.sequelize.modelManager.getModel(model.name).attributes;
				console.log('Processing foreign keys for ' + model.name);
				for (var columnName in columns) {
					var column = columns[columnName];
					if (column.references) {
						//if (column.references.key == column.references.model + '_id') {
						if (model.name == column.references.model) {
							console.log("  " + model.name + ' has one ' + column.fieldName.replace("_id", ""));
							model.belongsTo(model, { as: column.fieldName.replace("_id", ""), foreignKey: column.fieldName });
						} else {
							//console.log(model.name + ' has one ' + column.references.model);
							//console.log(column.references.model + ' has many ' + model.name);

							// Will add foreignKey to the right-value table
							//console.log("  " + model.name + " belongs to " + column.references.model + " on foreign key " + column.references.key);
							//model.belongsTo(instance.sequelize.models[column.references.model], { foreignKey: column.references.key });

							console.log("  " + column.references.model + " belongs to " + model.name + " on foreign key " + column.fieldName );
							instance.sequelize.models[column.references.model].belongsTo(model, { foreignKey: column.fieldName });

							console.log(model);

							//model.hasOne(instance.sequelize.models[column.references.model]);
							//instance.sequelize.models[column.references.model].hasMany(model);
						}
					}
				}
			}
```

**Comment by willemmulder** [comment]
Any news on this? Would be a great feature to have!

**Comment by a-gerold** [comment]
I started trying generating the associations at runtime, after importing my models, because my Database was already existing and I used sequelize-auto only for model generation.

Therefore i use the references property in my model, which was created by sequelize-auto.



```javascript
const database = new Sequelize(....)
...

database.modelManager.models.forEach(model => {
    Object.getOwnPropertyNames(model.attributes).forEach(attributeName => {
        if (model.attributes[attributeName].references) {
            console.log('-----------')
            console.log(attributeName + "->" +
                model.attributes[attributeName].references.model + "." +
                model.attributes[attributeName].references.key);

            var refModel = database.models[model.attributes[attributeName].references.model];
            model.belongsTo(refModel, {
                foreignKey: attributeName,
                as: attributeName.split('Id')[0]
            });

            console.log('mapping as: ' + attributeName.split('Id')[0] + ' on ' + attributeName)
            console.log('-----------')
        }
    })
})

```

I make use of "as", because in our Databse, all FK end with "Id".

I think it is possible to create also hasOne() in this Step by switching roles of model and refModel like so:
(untested)

```javascript
var refModel = database.models[model.attributes[attributeName].references.model];
refModel.hasOne(model, {
foreignKey: attributeName,
as: model.name
});

```

**Comment by victorlm125** [comment]
New news? I think that your idea is great but I think that you can't resolt many to many relations because you can't discover just analizing the model as now.

Have you found out a way to resolt many to many relations?

**Comment by sarikaya** [comment]
Is there any progress about the issue?

**Comment by victorlm125** [comment]
Certanly not. I still doing it manually. I have to try a way to automate but at the moment I have no time. The name of the relations table (in my case) are always with underscore (Ex: cars_users) so when I loop all tables I can distinguish them and créate many to many relations. Maybe it could be put on the road.

Regards,
Victor
Enviado desde Correo para Windows 10

De: sarikaya
Enviado: jueves, 11 de enero de 2018 19:58
Para: sequelize/sequelize-auto
CC: victorlm125; Comment
Asunto: Re: [sequelize/sequelize-auto] Generate model associations(belongsTo, hasMany, belongsToMany) from foreign keys (#34)

Is there any progress about the issue?

**Comment by zypA13510** [comment]
[The solution by @a-gerold](#issuecomment-331147595) is pretty good but I found a few issues when I run it:
1. `fs.readdirSync('./models')` somehow gives me an `Error: ENOENT: no such file or directory, scandir './models'`, changing it to `fs.readdirSync(__dirname + '/models')` worked for me.
2. `foreignKey: attributeName` should be changed to `foreignKey: model.attributes[attributeName].field` because, when using camel case, attributeName would be different from what's defined in the DB. Same with `var refModel = sequelize.models[model.attributes[attributeName].references.model]`.

So this is my final script after the said modification:
```js
const sequelize = new Sequelize(options)
// import all models
fs.readdirSync(__dirname + '/models').forEach(file => {
  sequelize.import('./models/' + file)
})
// setup belongsTo() based on references
sequelize.modelManager.models.forEach(model => {
  Object.getOwnPropertyNames(model.attributes).forEach(attributeName => {
    if (model.attributes[attributeName].references) {
      var refModel = sequelize.modelManager.models.find(
        testModel => testModel.tableName === model.attributes[attributeName].references.model
      )
      model.belongsTo(refModel, {
        foreignKey: model.attributes[attributeName].field,
        as: attributeName.split('Id')[0],
      })
    }
  })
})
```

By the way, for this to work you'll have to make sure all references end with "Id", fields like "createdBy" can easily get in the way here.

**Comment by smashmyster** [comment]
This worked for me with a few alterations. 
You just need to know where the models are stored 

In this method, the foreign key doesn't need to end with Id 
`const Sequelize = require('sequelize')

var db = {}

const sequelize = new Sequelize('testdata', 'root', '', {
    host: 'localhost',
    port: '3306',
    dialect: 'mysql',
    define: {
        freezeTableName: true,
    },
    pool: {
        max: 5,
        min: 0,
        acquire: 30000,
        idle: 10000,
    },
    // http://docs.sequelizejs.com/manual/tutorial/querying.html#operators
    operatorsAliases: false,
})
let  fs = require('fs');
let  files = fs.readdirSync('./app/models');
let models = []
files.forEach((file)=>{
    if(file.includes(".js")) {
        models.push(require('./models/' + file))
    }
})
// Initialize models
models.forEach(model => {
    const seqModel = model(sequelize, Sequelize)
    db[seqModel.name] = seqModel

})
Object.keys(db).forEach(modelName => {
    let model=db[modelName]
    let modelAttributes=model.rawAttributes
    let modelAttributeKeys=Object.keys(modelAttributes)
    modelAttributeKeys.forEach(attribute=>{
        if(modelAttributes[attribute].references){
            let linkData=modelAttributes[attribute].references
            let refModelName = Object.keys(db).find(
                model => model === linkData.model
            )
            if(refModelName){
                let refModel=db[refModelName]
                model.belongsTo(refModel,{
                    foreignKey:attribute,
                })
            }

        }
    })
})
db.sequelize = sequelize
db.Sequelize = Sequelize

module.exports = db
`

**Comment by troykelly** [comment]
My solution to auto generation of relationships is below, I'd love some feedback if I am doing something wrong - if not - maybe it could be a function in the core instead of being called like this?
If somebody knows an easier way to get the model names from the pluralised table names - I'd love to know!

```javascript
  const debug = require('debug')('db:autocreate');
  const tableModel = {};

  Object.getOwnPropertyNames(sequelize.models).forEach(function (modelName) {
    const tableName = sequelize.models[modelName].getTableName();
    tableModel[tableName] = sequelize.models[modelName];
  });

  Object.getOwnPropertyNames(sequelize.models).forEach(function (modelName) {
    const currentModel = sequelize.models[modelName];
    Object.getOwnPropertyNames(currentModel.rawAttributes).forEach(function (attributeName) {
      if (Object.prototype.hasOwnProperty.call(currentModel.rawAttributes[attributeName], "references") && Object.prototype.hasOwnProperty.call(currentModel.rawAttributes[attributeName].references, "model") && Object.prototype.hasOwnProperty.call(currentModel.rawAttributes[attributeName].references, "key")) {
        if (!(currentModel.rawAttributes[attributeName].references.model && currentModel.rawAttributes[attributeName].references.key)) {
          debug(`*SKIPPED* ${modelName} ${attributeName} references a model ${currentModel.rawAttributes[attributeName].references.model} with key ${currentModel.rawAttributes[attributeName].references.key}`);
          return;
        }

        debug(`${modelName} ${attributeName} references a model ${currentModel.rawAttributes[attributeName].references.model} with key ${currentModel.rawAttributes[attributeName].references.key}`);
        const referencedTable = tableModel[currentModel.rawAttributes[attributeName].references.model];

        currentModel.belongsTo(referencedTable, { foreignKey: attributeName });
        referencedTable.hasMany(currentModel, { foreignKey: attributeName });

      }
    })
  })

```

**Comment by smashmyster** [comment]
I ended up using this method. It is derived from the references in the file 
`models.forEach(model => {
    const seqModel = model(sequelize, Sequelize)
    db[seqModel.name] = seqModel

})
Object.keys(db).forEach(modelName => {
    let model = db[modelName]
    let modelAttributes = model.rawAttributes
    let modelAttributeKeys = Object.keys(modelAttributes)
    modelAttributeKeys.forEach(attribute => {
        if (modelAttributes[attribute].references) {
            let linkData = modelAttributes[attribute].references
            let refModelName = Object.keys(db).find(
                model => model === linkData.model
            )
            if (refModelName) {
                let refModel = db[refModelName]
                if (trans.associations[refModelName]) {
                    model.belongsTo(refModel, {
                        foreignKey: attribute,
                        onDelete: 'CASCADE',
                        as: trans.associations[refModelName]
                    })
                } else {
                    model.belongsTo(refModel, {
                        foreignKey: attribute,
                        onDelete: 'CASCADE',

                    })
                }
            }

        }
    })
})`

**Comment by smashmyster** [comment]
What is the problem with this method. It looks correct what error is it giving you?

**Comment by CodTicker** [comment]
I ran to this problem a few days ago and found that @troykelly solution works. 

However it does not take into account that you can have junction-tables and use the sequalize 
`Model.belongsToMany(OtherModel, { through: JunctionModel});`

I added my take on it and came up with the following: 

```
enum Junctions {
  user = 'user',
  roles = 'roles',
}

enum juntiontThrough {
  userroles = 'userroles',
}

interface JunctionObject {
  junctionBelongsTo?: string;
  juntiontThrough?: string;
}

const functionExecutor = (obj, value) => {
  return Object.prototype.hasOwnProperty.call(obj, value);
};

const checkRefrence = (obj, attr, value) => {
  return obj.rawAttributes[attr].references[value];
};


```
```
export const getJunction = (junc: Junctions): JunctionObject => {
  const junctions = {
    user: {
      junctionBelongsTo: 'roles',
      juntiontThrough: juntiontThrough.userroles,
    },
    roles: {
      junctionBelongsTo: 'user',
      juntiontThrough: juntiontThrough.userroles,
    },
  }[junc];

  if (!junctions) return {};

  return junctions;
};
```


```
```
my own take on @troykelly solution with added junction table code:

```
const models = Object.getOwnPropertyNames(db);

models.forEach(modelName => {
  const currentModel = db[modelName];
  const junction = getJunction(modelName as Junctions);

  if (!_.isEmpty(junction)) {
    currentModel.belongsToMany(db[junction.junctionBelongsTo], {
      through: db[junction.juntiontThrough],
    });
  }

  const attributes = Object.getOwnPropertyNames(currentModel.rawAttributes);

  attributes.forEach(attributeName => {
    if (
      functionExecutor(currentModel.rawAttributes[attributeName], 'references') &&
      functionExecutor(currentModel.rawAttributes[attributeName].references, 'model') &&
      functionExecutor(currentModel.rawAttributes[attributeName].references, 'key')
    ) {
      if (
        !(
          checkRefrence(currentModel, attributeName, 'model') &&
          checkRefrence(currentModel, attributeName, 'key')
        )
      ) {
        return;
      }


      if (!(modelName.toString() in juntiontThrough)) {
      const referencedTable =
        tableModel[currentModel.rawAttributes[attributeName].references.model];

        console.log(
          `${modelName} ${attributeName} references a model ${currentModel.rawAttributes[attributeName].references.model} with key ${currentModel.rawAttributes[attributeName].references.key}`
        );
        currentModel.belongsTo(referencedTable, { foreignKey: attributeName });
        referencedTable.hasMany(currentModel, { foreignKey: attributeName });
      }
    }
  });
});
```

By adding the above I could do this in my resolver: 

```
      const users = await db.user.findAll({
        include: [db.roles],
     });
```

instead of: 

```
      const users = await db.user.findAll({
        include: [
          {
            model: db.userroles,
            include: [
              {
                model: db.roles,
                attributes: ['Name'],
              },
            ],
          },
        ],
      });
```

This works for me but I might have missed something so if anyone has any thoughts please share.

**Comment by CodTicker** [comment]
A follow up to above: 

Currently you would have to define each many to many relation as a prop. 

I just changed  it to an array: 


```
export const getJunction = (junc: Junctions): JunctionObject => {
  const junctions = {
    user: {
      junctionBelongsTo: [{ key: Junctions.roles, value: juntiontThrough.userroles }], // add all junctions here
    },
    roles: {
      junctionBelongsTo: [{ key: Junctions.user, value: juntiontThrough.userroles }],// add all junctions here
    },
  }[junc];

  if (!junctions) return {};

  return junctions;
};
```

when executing: 

```
  if (!_.isEmpty(junction)) {
    // eslint-disable-next-line array-callback-return
    junction.junctionBelongsTo.reduce((key, value) => {
      currentModel.belongsToMany(db[value.key], {
        through: db[value.value],
      });
    }, {});
  }
```

**Comment by troykelly** [comment]
@smashmyster Sorry for delay responding - the code is working really well, I should have made that clear. My question/concern is that I don't understand why this isn't core functionality - am I doing something that I shouldn't by automating this?

@CodTicker great work! I have a tonne of junction tables (which is why I wrote the automation in the first place)
There must be a way to apply logic - ie `if there are two primary keys, and they both reference another table's primary key` ... something like that anyway.

**Comment by osvaldokalvaitir** [comment]
Is there any progress about the issue?

**Comment by steveschmitt** [comment]
No, but I hope to work on it soon.  Very busy right now.

**Comment by odravison** [comment]
I'm so excited to see this happening!!!

**Comment by ericfr** [comment]
from @troykelly answer, i had to get the referencedTable like this:
const referencedTable =
        tableModel[
          currentModel.rawAttributes[attributeName].references.model.tableName
        ];

notice the extra .tableName

**Comment by steveschmitt** [comment]
Added `belongsTo/hasOne/hasMany` to the generated `initModels` function in release **0.7.0**.

No support yet for `belongsToMany`.  I need the right heuristic for determining what should be junction table.  But I wanted to get this out there so you can try it.

Please try the latest release and see how it works for you.

**Comment by steveschmitt** [comment]
I added `belongsToMany` in release **0.7.1** (and also fixed a bug with the foreign keys in `belongsTo`.  

SequelizeAuto will generate the `belongsToMany` if the junction entity has two foreign keys that are also part of the primary key.

**Comment by osvaldokalvaitir** [comment]
Thank you Steve! this will help me a lot, i will perform some tests here =)

**Comment by KjellVermeersch** [comment]
I have noticed that the current behaviour
> SequelizeAuto will generate the belongsToMany if the junction entity has two foreign keys that are also part of the primary key.

will recognize many-to-many relation ships as a charm (for most cases). But, I have a one-to-one relationship with two foreign keys as primary keys in both tables. This means that instead of a hasOne association, it generates two belongsToMany associations to itself. This isn't correctly (and Sequelize will throw an error because it expects aliases when using an association to itself).

This is definitly an edge case, so I am just informing you about it. If you need me to add some more details, let me know.

**Comment by steveschmitt** [comment]
@KjellVermeersch I'm not sure I'll fix it right away, but please let me know the details, i.e. the table definitions, the actual output, and the expected output.  Maybe you should open a separate issue for this.

**Comment by steveschmitt** [comment]
I'm going to close this issue, since the core behavior is all implemented as of 0.7.6.

@KjellVermeersch Please file a separate issue describing the bug that you found.

Also note that there is a separate issue #466 regarding multiple foreign keys.
